// views/m/user/LoginForm/index.ts
import * as tomato from "@po-to/tomato";
import * as project from "views/global/common/Project";
import * as Vue from "vue";
import * as model from "views/global/common/Model";
import * as api from "views/global/common/API";
import * as funs from 'views/global/common/Funs';
import * as tdom from "@po-to/tomato-jquery";
import VerificationCode = require('views/m/user/VerificationCode');
import code = require('views/m/user/Code');

class VPresenter extends project.VPresenter {
    /**模块 */
    private LoginForm_vue: any;
    /**手机号 */
    private _phone: JQuery;
    /**验证码 */
    private _code: JQuery;
    /**有项目未完成时提示 */
    private _inquiryTips: JQuery;
    /**手机号状态图标 */
    private _phoneCheck: JQuery;
    /**错误提示 */
    private _codeCheck: JQuery;
    /**登录按钮 */
    private _login: JQuery;
    /**获取验证码按钮 */
    private _getCaptcha: JQuery;
    /**验证码是否可以再次发送 */
    private _flag: boolean;
    /**登录判断是否新用户 */
    private _isCallPublish: boolean;

    constructor(view: tomato.VPView, parent?: tomato.VPresenter, vpid?: string) {
        super(view, parent, vpid);
        var vueDom = this.find("#LoginForm_vue").get(0);
        var user = model.globalData.user;
        let LoginForm_vue: any = this.LoginForm_vue = new Vue({
            el: vueDom,
            data: {
                user: user,
                code: '',
                message: '',
                showIcon: false,
                showError: false,
                seen: false,
                isActive: false,
                clicked: true
            },
            watch: {
                'user.phone': "checkInfo",
                code: "checkInfo",
            },
            methods: {
                //检验必填项
                checkInfo: function () {
                    if (LoginForm_vue.user.phone && LoginForm_vue.code) {
                        LoginForm_vue.clicked = false;
                    } else {
                        LoginForm_vue.clicked = true;
                    }
                }
            }
        });
        this._phone = this.find(".phone");
        this._code = this.find(".code");
        this._inquiryTips = this.find(".inquiry-tips");
        this._phoneCheck = this.find(".phone-check");
        this._codeCheck = this.find(".code-check");
        this._login = this.find("#login_button");
        this._getCaptcha = this.find(".getcode");
        this._flag = true;
        this._isCallPublish = false;
        let that = this;
        //重新编辑手机号时，错误提示取消
        this._phone.bind("input propertychange", function () {
            LoginForm_vue.showIcon = false;
            LoginForm_vue.showError = false;
            that.find(".promptBar").addClass("hide");
        });

        this._watchEvent();

    }

    //检测手机号
    private _check_phone(usrName: string): boolean {
        if (!usrName) {
            // this.LoginForm_vue.showIcon = true;
            this.LoginForm_vue.isActive = false;
            // this.LoginForm_vue.showError = true;
            // this.LoginForm_vue.message = "手机号不能为空";
            this.find(".promptBar").removeClass("hide");
            this.find(".promptBar").text('手机号不能为空');
            return false;
        } else if (!this._checkPhone(usrName)) {
            // this.LoginForm_vue.showIcon = true;
            this.LoginForm_vue.isActive = false;
            // this.LoginForm_vue.showError = true;
            // this.LoginForm_vue.message = "请填写正确的手机号";
            this.find(".promptBar").removeClass("hide");
            this.find(".promptBar").text('请填写正确的手机号');
            return false;
        } else {
            this.LoginForm_vue.showIcon = true;
            this.LoginForm_vue.isActive = true;
            this.LoginForm_vue.showError = false;
            return true;
        }
    }
    //正则检测手机号
    private _checkPhone(phone: string): boolean {
        if (!(/^1(3|4|5|7|8)\d{9}$/.test(phone))) {
            return false;
        }
        return true;
    }

    //提交表单
    _evt_submitForm(data, target) {
        var that = this;
        //得有一个接口去验证手机号是不是工程商

        if (that._check_phone(that.LoginForm_vue.user.mobile)) {
            //
            api.GetInfoByMobile(that.LoginForm_vue.user.mobile, function (json) {
                //success
                // window.location.href = '/Index';
                if (json.user && json.user.type == 100) {
                    //是工程商
                    //请求验证码发送接口
                    tomato.getVPresenter<code>(funs.moduleToUrl('m/user/Code'), (code) => {
                        code._evt_getCode();
                        code.setCallBack(function () { window.location.href = "/index.html" });
                        funs.loadPage(code);
                    });
                } else {
                    //不是工程商
                    tomato.getVPresenter<tomato.VPresenter>(funs.moduleToUrl('m/joinUs/JoinSupplierPop'), (JoinSupplierPop) => {
                        //您不是工程商，需要立即加盟提示
                        tdom.open(JoinSupplierPop, tdom.DialogTarget.Blank, {
                            size: { width: '90%', height: tomato.DialogSize.Content },
                            effect: tdom.DialogEffect.scale,
                            masked: true
                        });
                    });
                }
            }, function () {
                //fail
                tomato.getVPresenter<tomato.VPresenter>(funs.moduleToUrl('m/joinUs/JoinSupplierPop'), (JoinSupplierPop) => {
                    //您不是工程商，需要立即加盟提示
                    tdom.open(JoinSupplierPop, tdom.DialogTarget.Blank, {
                        size: { width: '90%', height: tomato.DialogSize.Content },
                        effect: tdom.DialogEffect.scale,
                        masked: true
                    });
                });
            })

        }

        return false;
    };
    /**返回按钮 */
    _evt_goback(n?: number) {
        if (model.globalData.user.user_type == 0) {
            //未登录，证明是从HomePage进来的，允许返回
            n = n || -1;
            funs.viewHistory.go(n);
        }else{
            //登录了，出现此页代表不是工程商，必须加盟才可以进入，如果点击返回，则退出App

        }

    }
}
export = VPresenter;
