// views/m/joinUs/joinSupplierBasicInfo/index.ts
import * as tomato from "@po-to/tomato";
import * as tdom from "@po-to/tomato-jquery";
import * as project from "views/global/common/Project";
import * as funs from "views/global/common/Funs";
import * as model from "views/global/common/Model";
import * as api from "views/global/common/API";
import * as Vue from "vue";

class VPresenter extends project.VPresenter {
    /**模块 */
    public BasicInfo_vue: any;
    /**user */
    private user;
    private errorTip;
    constructor(view: tomato.VPView, parent?: tomato.VPresenter, vpid?: string) {
        super(view, parent, vpid);

        var vueDom = this.find("#BasicInfo_vue").get(0);
        var user = this.user = model.globalData.user;

        let that = this;

        let BasicInfo_vue: any = this.BasicInfo_vue = new Vue({
            el: vueDom,
            data: {
                user: user
            },
        });
        this.errorTip = this.find(".errorTip");
        $('input').bind("input propertychange", function () {
            // that.errorTip.addClass("hide");
            that.find(".promptBar").addClass('hide');
        });
        this._watchEvent();
        this._els = this._getElements();
    }
    init() {
        if (model.globalData.user.user_type != 0) {
            //已经登陆
            this.find(".phone").attr("disabled", "true").addClass('disabled');
        }
    }
    /**昵称校验 */
    private checkNick(info, dom, tip) {
        var reg = /^[\u4e00-\u9faf]+$/
        if (info == "") {
            // $('.' + dom + 'Tip').removeClass("hide");
            // $('.' + dom + 'Tip').find('.error-data').text(tip + '不能为空！')
            this.find(".promptBar").removeClass('hide');
            this.find(".promptBar").text(tip + '不能为空！');
            return false;
        } else if (!reg.test(info)) {
            // $('.' + dom + 'Tip').removeClass("hide");
            // $('.' + dom + 'Tip').find('.error-data').text('请输入正确的' + tip);
            this.find(".promptBar").removeClass('hide');
            this.find(".promptBar").text('请输入正确的' + tip);
            return false;
        } else {
            return true;
        }
    }
    /**手机号 */
    private checkPhone(info, dom, tip) {
        var reg = /^1(3|4|5|7|8)\d{9}$/
        if (info == "") {
            // $('.' + dom + 'Tip').removeClass("hide");
            // $('.' + dom + 'Tip').find('.error-data').text(tip + '不能为空！')
            this.find(".promptBar").removeClass('hide');
            this.find(".promptBar").text(tip + '不能为空！');
            return false;
        } else if (!reg.test(info)) {
            // $('.' + dom + 'Tip').removeClass("hide");
            // $('.' + dom + 'Tip').find('.error-data').text('请输入正确的' + tip);
            this.find(".promptBar").removeClass('hide');
            this.find(".promptBar").text('请输入正确的' + tip);
            return false;
        } else {
            return true;
        }
    }
    /**下一步 */
    _evt_joinNext() {
        if (this.checkNick(this.find('.nick').val(), 'nick', "昵称") && this.checkPhone(this.find('.phone').val(), 'phone', "手机号")) {
            tomato.getVPresenter<tomato.VPresenter>(funs.moduleToUrl('m/joinUs/joinSupplierCompanyInfo'), (joinSupplierCompanyInfo) => {
                //弹出详情页
                funs.loadPage(joinSupplierCompanyInfo);
            })
        }
    }
    _afterInstallTo() {
        //加载模块时
        // this.BasicInfo_vue.checkData();
        // this.BasicInfo_vue.nickError = false;
        // this.BasicInfo_vue.phoneError = false;
    }
    _afterUninstallTo() {
        //退出时
        // this.getGlobalData();
    }
    setGlobalData() {
        //this.vueData => globalData
        // $.extend(true, model.globalData.user, this.joinBasicData_vue.user);

    } 
    getGlobalData() {
        //globalData => this.vueData
        // this.joinBasicData_vue.user = funs.deepCopy(model.globalData.user);
    }
    /**返回按钮 */
    _evt_goback(n?: number) {
        if (model.globalData.user.user_type == 0) {
            //未登录，证明是从HomePage进来的，允许返回
            n = n || -1;
            funs.viewHistory.go(n);
        }else{
            //登录了，出现此页代表不是工程商，必须加盟才可以进入，如果点击返回，则退出App

        }

    }
}
export = VPresenter;
