// views/m/joinUs/joinSupplierCompanyInfo/index.ts
import * as tomato from "@po-to/tomato";
import * as tdom from "@po-to/tomato-jquery";
import * as project from "views/global/common/Project";
import * as funs from "views/global/common/Funs";
import * as model from "views/global/common/Model";
import * as api from "views/global/common/API";
import * as Vue from "vue";
import VerificationCode = require('views/m/user/VerificationCode');
import code = require('views/m/user/Code');

class VPresenter extends project.VPresenter {
    /**模块 */
    public CompanyInfo_vue: any;
    /**user */
    private user;
    private errorTip;
    constructor(view: tomato.VPView, parent?: tomato.VPresenter, vpid?: string) {
        super(view, parent, vpid);
        var vueDom = this.find("#CompanyInfo_vue").get(0);
        var user = this.user = model.globalData.user;

        let that = this;

        let CompanyInfo_vue: any = this.CompanyInfo_vue = new Vue({
            el: vueDom,
            data: {
                user: user
            },
        });
        this.errorTip = this.find(".errorTip");
        $('input').bind("input propertychange", function () {
            that.errorTip.addClass("hide");
        });
        this._watchEvent();
        this._els = this._getElements();
    }
    /**昵称校验 */
    private checkNick(info, dom, tip) {
        var reg = /^[\u4e00-\u9faf]+$/
        if (info == "") {
            $('.' + dom + 'Tip').removeClass("hide");
            $('.' + dom + 'Tip').find('.error-data').text(tip + '不能为空！')
            return false;
        } else if (!reg.test(info)) {
            $('.' + dom + 'Tip').removeClass("hide");
            $('.' + dom + 'Tip').find('.error-data').text('请输入正确的' + tip);
            return false;
        } else {
            return true;
        }
    }
    /**手机号 */
    private checkPhone(info, dom, tip) {
        var reg = /^1(3|4|5|7|8)\d{9}$/
        if (info == "") {
            $('.' + dom + 'Tip').removeClass("hide");
            $('.' + dom + 'Tip').find('.error-data').text(tip + '不能为空！')
            return false;
        } else if (!reg.test(info)) {
            $('.' + dom + 'Tip').removeClass("hide");
            $('.' + dom + 'Tip').find('.error-data').text('请输入正确的' + tip);
            return false;
        } else {
            return true;
        }
    }
    /**提交 */
    _evt_joinSubmit() {
        var that = this;
        let nickname = model.globalData.user.nick;
        let mobile = model.globalData.user.mobile;
        let gender = model.globalData.user.sex;
        let prov = model.globalData.province.province_name;
        let city = model.globalData.city.city_name;
        let addr = model.globalData.user.address;
        let com = model.globalData.user.company_name;
        let pos = model.globalData.user.position;
        let userInfo = {
            nickname: nickname,
            mobile: mobile,
            gender: gender,
            prov: prov,
            city: city,
            addr: addr,
            com: com,
            pos: pos
        };
        /**加盟工程商接口 */
        api.SupplierJoin(userInfo, function () {
            //成功，刷新
            window.location.href = '/index.html';
        }, function (error: tomato.PError) {
            if (error.name == '401') {
                //未登录，去执行登录
                //弹出验证码

                tomato.getVPresenter<code>(funs.moduleToUrl('m/user/Code'), (code) => {
                    code.setCallBack(function () { that._evt_joinSubmit() });
                    tdom.open(code, tdom.DialogTarget.Blank, {
                        size: { width: '100%', height: '100%' },
                        effect: tdom.DialogEffect.slideLeft,
                        masked: true,
                    });
                });
            } else if (error.name == '41009003') {
                //已是工程商
                funs.tip(error.note, "", "fail");
                setTimeout(function () {
                    window.location.href = '/index.html';
                }, 3020);
            } else {
                funs.tip(error.name, error.note, "fail");
            }
        });
    }
}
export = VPresenter;
