// views/m/common/NewFooter/index.ts
import * as tomato from "@po-to/tomato";
import * as tdom from "@po-to/tomato-jquery";
import * as project from "views/global/common/Project";
import * as funs from "views/global/common/Funs";
import * as model from "views/global/common/Model";
import * as api from "views/global/common/API";
import * as Vue from "vue";

class VPresenter extends project.VPresenter {

    public rightDialog: tdom.CommonDialog;
    private _sign: string;
    private _saveTarget: any;
    private _saveHit: any;
    constructor(view: tomato.VPView, parent?: tomato.VPresenter, vpid?: string) {
        super(view, parent, vpid);
        this._sign = "";
        this._saveTarget = this.find('.active');
        var that = this;
        let vueDom = this.find("#news_vue").get(0);
        let user = model.globalData.user;
        let unread = model.globalData.unread_num;
        new Vue({
            el: vueDom,
            data: {
                isActive: false,
                unread: unread
            },
            computed: {
                unreadCom: function () {
                    if ((this as any).unread.private_unread > 9) {
                        return '…';
                    }
                    return (this as any).unread.private_unread;
                }
            }
        });
        this.rightDialog = new tdom.CommonDialog({
            className: "rightMenu",
            effect: tdom.DialogEffect.slideLeft,
            masked: true,
            size: { width: 300, height: tomato.DialogSize.Full },
            position: { x: tomato.DialogPosition.Right, y: tomato.DialogPosition.Top },
            bodyEffect: tdom.TurnEffect.slid,
            asideEffect: tdom.TurnEffect.slid,
        });
        this.rightDialog.close = function (): boolean {
            that._evt_openBlock("", that._saveTarget, that._saveHit)
            return tdom.CommonDialog.prototype.close.call(this, false);
        }
        tomato.application.appendChild(this.rightDialog);
        this._watchEvent();
        this._els = this._getElements();
    }
    _evt_openBlock(data, target, hit) {
        var that = this;
        var name = target.innerHTML;
        var sign = $(target).attr("sign");
        if (that._sign == sign) {
            return;
        }
        /**移除所有active */
        that.classReset();
        $(target).addClass('active');
        $(target).children("span:first").attr('class', $(target).children("span:first").attr('class') + '1');
        if (sign == "Index") {
            that._saveTarget = target;
            that._saveHit = hit;
            //发现
            tomato.getVPresenter<tomato.VPresenter>(funs.moduleToUrl('m/Index'), (Index) => {
                funs.loadPage(Index);
            })
        } else if (sign == "Message") {
            that._saveTarget = target;
            that._saveHit = hit;
            //消息
            tomato.getVPresenter<tomato.VPresenter>(funs.moduleToUrl('m/project/JoinedProjects'), (JoinedProjects) => {
                funs.loadPage(JoinedProjects);
            })
        }
        else if (sign == "UserCenter") {
            //个人中心
            tomato.getVPresenter<tomato.VPresenter>(funs.moduleToUrl('m/user/UserCenter'), (userCenter) => {
                this.rightDialog.appendChild(userCenter);
                this.rightDialog.focus();
            })
        }

    }
    /**移除所有active */
    classReset() {
        this.find('.foot-block').removeClass('active');
        this.find('.faxian').children("span:first").attr('class', 'icon-faxian');
        this.find('.news').children("span:first").attr('class', 'icon-news');
        this.find('.me').children("span:first").attr('class', 'icon-me');
    }
}
export = VPresenter;
