// views/m/project/PersonalHomePage/index.ts
import * as tomato from "@po-to/tomato";
import * as tdom from "@po-to/tomato-jquery";
import * as project from "views/global/common/Project";
import * as funs from "views/global/common/Funs";
import * as model from "views/global/common/Model";
import * as api from "views/global/common/API";
import * as demand from 'nodejs/global/model/demand';
import * as Vue from 'vue';
import DemandDetail = require('views/m/DemandDetail');
import UpdateSupplierInfo = require('views/m/user/UpdateSupplierInfo');

class VPresenter extends project.VPresenter {
    private nick;
    private project_list;
    private current_demand;
    private PersonalHomePage_vue;
    private project_list_array;
    private flag_html: string;
    constructor(view: tomato.VPView, parent?: tomato.VPresenter, vpid?: string) {
        super(view, parent, vpid);
        this.nick = model.globalData.current_demand.user.nick.substring(0, 1);
        this.project_list = model.globalData.currUserDemandList;
        this.current_demand = model.globalData.current_demand;
        var vueDom = this.find(".PersonalHomePage_vue").get(0);
        this.project_list_array = new Array();
        this.flag_html = "";
        var that = this;
        that.getDemandList();
        let PersonalHomePage_vue: any = window['aaa'] = that.PersonalHomePage_vue = new Vue({
            el: vueDom,
            data: {
                project_list: that.project_list,
                current_demand: that.current_demand,
                currentUserName: that.nick,
            },
            methods: {
                updateTime: function (startTime) {
                    startTime = startTime + '';
                    var text;
                    // var stime = new Date(startTime).getTime();//转换
                    var stime = startTime * 1000;
                    var date = new Date().getTime();
                    var gapMin = Math.floor((date - stime) / 1000 / 60);
                    if (gapMin <= 0) {
                        //防止偶尔出现的js取时间不准确
                        gapMin = 1;
                    }
                    var gapHour = Math.floor((date - stime) / 1000 / 60 / 60);

                    if (gapHour <= 24) {
                        text = gapHour + '小时前';
                        if (gapMin < 60) {
                            text = gapMin + '分钟前';
                        }
                    } else if (gapHour <= 24 * 30) {
                        gapHour = Math.floor(gapHour / 24);
                        text = gapHour + '天前'
                    } else if (gapHour <= 24 * 30 * 12) {
                        gapHour = Math.floor(gapHour / 24 / 30);
                        text = gapHour + '个月前'
                    } else {
                        //防止意外情况，暂时如此处理
                        text = '刚刚'
                    }
                    return text;
                },
                iJoined: function (list: any[]) {
                    return list.some(function (item) {
                        return item.id == model.globalData.user.user_id;
                    })
                }
            },

        })

        this._watchEvent();
        this._els = this._getElements();
    }
    private getDemandList() {
        var uid = model.globalData.current_demand.user.id;
        var that = this;
        if (uid) {
            api.GetUserDemandList({ uid: uid }, function (json) {
                //成功取回
                model.globalData.currUserDemandList = json;
                var length = json.length;
                that.project_list_array = [];
                for (var i = 0; i < length; i++) {
                    if (json[i].id != model.globalData.current_demand.id) {
                        that.project_list_array.push(json[i]);
                    }
                }
                that.PersonalHomePage_vue.project_list = that.project_list_array;
                that.find(".customer-demand-list").removeClass('hide');
            }, function (error) { })
        }

    }
    /**获取单个项目的详细信息 */
    getProjectData(did: string): demand.DemandInfo | undefined {
        var that = this;
        for (let a of model.globalData.currUserDemandList) {
            if (a.id == did) {
                return a
            }
        }
    }
    /**项目的点击事件 */
    _evt_projectDetail(id: string) {
        var pro_info = this.getProjectData(id);
        if (pro_info) {
            model.globalData.current_demand = pro_info;
            var flag = model.globalData.demand.id == model.globalData.current_demand.id;
            model.globalData.currIsSelfDemand = flag;
            tomato.getVPresenter<DemandDetail>(funs.moduleToUrl('m/DemandDetail'), (DemandDetail) => {
                //弹出详情页
                if (DemandDetail.vpid.indexOf('demand=') > 0) {
                    DemandDetail.vpid = DemandDetail.vpid.replace(/\?(demand=\d*)?/, '?demand=' + (pro_info as any).id);
                } else {
                    DemandDetail.vpid += '?demand=' + (pro_info as any).id;
                }
                funs.loadPage(DemandDetail)
            })
        }
    }
    /**=======================点赞---有用  开始=====================*/
    _evt_thumbsUp(id: string, data, target, hit) {
        var demand_id = id;
        //当前id项目的所有信息
        var pro_info = this.getProjectData(demand_id);
        //修改选中类型
        var that = this;
        var name = data.innerHTML;
        var flag = parseInt($(data).attr("flag"));
        if (!pro_info) {
            return;
        }
        var voteInfo = {
            id: demand_id,
            type: flag//投票类型，1：顶；2：踩；
        };
        if (pro_info.vote_type == 0) {
            //顶
            api.Vote(voteInfo, function (json) {
                that.resetDemand(json);

            }, function () {
                //fail
            })
        }
        if (pro_info.vote_type == flag) {
            //取消
            api.Unvote(voteInfo, function (json) {
                that.resetDemand(json);
            }, function () {
                //fail
            })
        }
    }
    //
    resetDemand(json) {
        if (model.globalData.current_demand.id == json.id) {
            model.globalData.current_demand.vote_up = json.vote_up;
            model.globalData.current_demand.vote_down = json.vote_down;
            model.globalData.current_demand.vote_type = json.vote_type;
        }
        for (let a of model.globalData.currUserDemandList) {
            if (a.id == json.id) {
                a.vote_up = json.vote_up;
                a.vote_down = json.vote_down;
                a.vote_type = json.vote_type;
            }
        }
    }
    /**=======================点赞---有用  结束=====================*/
    /**关闭气泡 */
    _evt_TipHide() {
        this.find('.foot-botton .tip').addClass('hide');
    }
    /**打开聊天室 */
    _evt_onlineChat(id: string) {
        var user_id = id;
        tomato.getVPresenter<tomato.VPresenter>(funs.moduleToUrl('m/ChatRoom'), (ChatRoom) => {
            if (ChatRoom.vpid.indexOf('uid=') > 0) {
                ChatRoom.vpid = ChatRoom.vpid.replace(/\?(uid=\d*)?/, '?uid=' + user_id);
            } else {
                ChatRoom.vpid += '?uid=' + user_id;
            }
            funs.loadPage(ChatRoom);
        });
    }
    /*================抢单逻辑－start===============*/
    _evt_grab() {
        var id = model.globalData.current_demand.id;

        var pro_info: any = this.getProjectData(id);
        if (pro_info) {
            model.globalData.current_demand = pro_info;
        }
        var that = this;
        let userInfo = model.globalData.user;
        let demandInfo = model.globalData.current_demand;
        if (userInfo.user_type == 100) {
            if (userInfo.avatar) {
                /** 暂时不需要工程商缴费，所以，此处正常逻辑先屏蔽 */
                // tomato.getVPresenter<tomato.VPresenter>(funs.moduleToUrl('m/grab/PayADepositModal'), (PayADepositModal) => {
                //     tdom.open(PayADepositModal, tdom.DialogTarget.Blank, {
                //         size: { width: '90%', height: tomato.DialogSize.Content },
                //         masked: true,
                //     });
                // });
                /** 暂时不需要工程商缴费，所以，此处正常逻辑先屏蔽 */

                // 临时逻辑－ start- /
                //抢单
                api.grabDemand({ "demand_id": demandInfo.id }, function (json) {
                    // 更新项目信息
                    var curr_pro = model.globalData.current_demand;
                    curr_pro.supplier_partic_status = json.demand_info.supplier_partic_status;
                    curr_pro.supplier_list = json.demand_info.partic_supplier_list;
                    curr_pro.bid_cnt = json.demand_info.demand.bid_cnt;
                    curr_pro.status = json.demand_info.demand.status;
                    curr_pro.user.phone = json.demand_info.user.mobile;

                    var pro_list = model.globalData.project_list;
                    // 更新项目列表中此项目的信息
                    for (var i = 0; i < pro_list.length; i++) {
                        if (curr_pro.id == pro_list[i].id) {
                            pro_list[i] = curr_pro;
                            break;
                        }
                    }
                    // 更新参与项目列表信息
                    let joined_list = model.globalData.joined_list;
                    let hasPro = false;
                    for (var i = 0; i < joined_list.length; i++) {
                        if (curr_pro.id == joined_list[i]) {
                            joined_list[i] = $.extend(joined_list[i], curr_pro);
                            hasPro = true;
                        }
                    }
                    if (!hasPro) {
                        joined_list.push(curr_pro);
                    }
                    // 自动拨打电话
                    window.location.href = 'tel:' + curr_pro.user.phone;
                },
                    function (data: tomato.PError) {
                        //如果请求失败,刷新当前
                        funs.tip('请求失败，请重试', data.note, 'fail');
                    })
                // 临时逻辑－ end- /
            } else {
                //弹出完善个人资料信息弹窗
                tomato.getVPresenter<UpdateSupplierInfo>(funs.moduleToUrl('m/user/UpdateSupplierInfo'), (UpdateSupplierInfo) => {
                    UpdateSupplierInfo.setType('1');
                    tdom.open(UpdateSupplierInfo, tdom.DialogTarget.Blank, {
                        size: { width: '90%', height: tomato.DialogSize.Content },
                        masked: true,
                    });
                });
            }
        } else {
            //当前登录信息与前端逻辑不合，直接刷新
            window.location.reload();
        }
    }
    /*================抢单逻辑－end===============*/
    _afterInstallTo() {
        //加载模块时
        this.PersonalHomePage_vue.currentUserName = model.globalData.current_demand.user.nick.substring(0, 1);
        this.PersonalHomePage_vue.current_demand = model.globalData.current_demand;
        this.getDemandList();
    }
    _afterUninstallTo() {
        //退出时
        // this.getGlobalData();
    }
}
export = VPresenter;
