// views/m/user/Code/index.ts
import * as tomato from "@po-to/tomato";
import * as project from "views/global/common/Project";
import * as Vue from "vue";
import * as model from "views/global/common/Model";
import * as api from "views/global/common/API";
import * as funs from "views/global/common/Funs";

class VPresenter extends project.VPresenter {
    /**模块 */
    private Code_vue: any;
    /**demand_id */
    private demand_id;
    /**验证码倒计时 */
    private codeGain;
    /**span */
    private codeParent;
    private _call_back_function;
    private _code;
    private _timer;
    constructor(view: tomato.VPView, parent?: tomato.VPresenter, vpid?: string) {
        super(view, parent, vpid);
        window['deviceData'] = JSON.parse(localStorage.getItem("deviceData") + "");
        var vueDom = this.find("#Code_vue").get(0);
        var user = model.globalData.user;
        var demand = model.globalData.demand;
        let Code_vue: any = this.Code_vue = new Vue({
            el: vueDom,
            data: {
                user: user,
                demand: demand,
                data: '59s后重新获取验证码',
                code: '',
            },
            watch: {
                'code': "checkData",
            },
            methods: {
                checkData: function () {
                    Code_vue.code = Code_vue.code.replace(/[^\d]/g, '');
                }
            }
        });

        this.codeGain = this.find(".c-code-gain");
        this.codeParent = this.find(".c-code-put");
        this.find('.to-next').attr("disabled", "true").addClass('disabled')
        this.getCode();
        this.inputFocus();

        this._watchEvent();
        // this.addListener(tomato.DialogEvent.Closed, function () {
        //     funs.viewHistory.go(-1);
        // })
    }
    setCallBack(setCallBack) {
        this._call_back_function = setCallBack;
    }
    private getCode = function () {
        var that = this;
        that.codeCountDown();
    }

    private codeCountDown = function () {
        var that = this;
        that.codeGain.html('60s后重新获取验证码').prop('disabled', true);
        var wait = 60;
        clearInterval(that._timer);
        that._timer =null;
        that._timer = setInterval(function () {
            wait--;
            that.codeGain.html(wait + 's后重新获取验证码');
            if (wait < 1) {
                clearInterval(that._timer);
                that._timer = null;
                that.codeGain.html('重新获取验证码').prop('disabled', false).addClass('code-blue');
            }
        }, 1000)
    }

    public _evt_getCode = function () {
        var that = this;
        var phone = model.globalData.user.mobile;
        //a02发送手机验证码接口
        api.SendSmsCaptcha(phone, function (json) {
            //倒计时
            that.codeCountDown();
        }, function (error: tomato.PError) {
            // that.find(".promptBar").removeClass('hide');
            // that.find(".promptBar").text(error.note);
            // setTimeout(function () {
            //     that.find(".promptBar").addClass('hide');
            // }, 3000);
        });
    }

    private inputFocus = function () {
        var that = this;

        this.codeParent.bind('input propertychange', 'input', function (event) {
            var sum = $(this).val().trim();
            if (!sum) {
                $(this).val('')
            } else {
                $(this).val($(this).val().trim());
            }
            if (4 == sum.length) {
                // $(this).attr('contenteditable', 'false');
                console.log(sum + '4位数字，尝试校验...');
                $(this).blur();
                that.find('.to-next').attr("disabled", false).removeClass('disabled')
                that._code = sum;
            }
        });
    }
    /**登录 */
    _evt_toLogin() {
        var that = this;
        var phone = model.globalData.user.mobile;
        var code = this._code;
        var push_reg_id = "";//window['deviceData'].registerId;
        if (window['deviceData'] && window['deviceData'].register_id) {
            push_reg_id = window['deviceData'].register_id + "";
        }
        if (code) {
            api.Login(phone, code, push_reg_id, function (data: any) {
                // alert("push_reg_id==" + push_reg_id)
                that._call_back_function();
            }, function (error: tomato.PError) {
                funs.tip("验证码有误，请重新输入", "", "fail");
            });
        }
    }
    _afterInstallTo() {
        //加载模块时
        $(document).ready(function () { $("input[name=focus]").focus(); });
        // this.Code_vue.user.mobile=model.globalData.user.mobile;
        this.find(".mobile").text(model.globalData.user.mobile);
    }
}
export = VPresenter;
