// views/m/user/VerificationCode/index.ts
import * as tomato from "@po-to/tomato";
import * as project from "views/global/common/Project";
import * as Vue from "vue";
import * as model from "views/global/common/Model";
import * as api from "views/global/common/API";
import * as funs from "views/global/common/Funs";

class VPresenter extends project.VPresenter {
    /**模块 */
    private Code_vue: any;
    /**手机号 */
    private phone;
    /**demand_id */
    private demand_id;
    /**验证码倒计时 */
    private codeGain;
    /**span */
    private codeParent;
    private _code;
    private _call_back_function;
    private _flag;

    constructor(view: tomato.VPView, parent?: tomato.VPresenter, vpid?: string) {
        super(view, parent, vpid);

        var vueDom = this.find("#Code_vue").get(0);
        var user = model.globalData.user;
        var demand = model.globalData.demand;
        this._code = "";
        this._flag = true;
        var that = this;
        let Code_vue: any = this.Code_vue = new Vue({
            el: vueDom,
            data: {
                user: user,
                data: '60s后可重新发送',
            }
        });
        this.phone = Code_vue.user.phone;
        this.codeGain = this.find(".c-code-gain");
        this.codeParent = this.find(".c-code-put");
        this.find('.to-next').attr("disabled", "true").addClass('disabled')
        this.find('.now').keyup(function () {
            if ($(this).index() < 5) {
                if ($(this).val().trim()) {
                    $(this).next('input').focus();
                } else {
                    $(this).val('');
                }

            }
        });
        this.getCode();
        this.inputFocus();


        this._watchEvent();
        // this.addListener(tomato.DialogEvent.Closed, function () {
        //     funs.viewHistory.go(-1);
        // })
    }
    setDemandId(demand_id) {
        this.demand_id = demand_id;
    }
    /**回调函数 */
    setCallBack(setCallBack) {
        this._call_back_function = setCallBack;
    }
    private getCode = function () {
        var that = this;
        that.codeCountDown();
    }

    private codeCountDown = function () {
        var that = this;
        that._flag = false;
        that.codeGain.html('60s后可重新发送').prop('disabled', true);
        var wait = 60;
        var timer = setInterval(function () {
            wait--;
            that.codeGain.html(wait + 's后可重新发送');
            if (wait < 1) {
                clearInterval(timer);
                that._flag = true;
                this.timer = null;
                that.codeGain.html('重新获取验证码').prop('disabled', false).addClass('code-blue');
            }
        }, 1000)
    }
    private inputFocus = function () {
        var that = this;
        this.find('.now').bind('input propertychange', 'input', function (event) {
            that.find('.to-next').attr("disabled", true).addClass('disabled')
            var sum = "";
            that.find('.now').each(function () {
                sum += $(this).val().trim();
            });
            if (4 == sum.length) {
                console.log(sum + '4位数字，尝试校验...');
                that._code = sum;
                that.find('.to-next').attr("disabled", false).removeClass('disabled')
            }
        });
    }
    /**发送验证码 */
    public _evt_getCode = function () {
        var that = this;
        var phone = model.globalData.user.mobile;
        //a02发送手机验证码接口
        api.SendSmsCaptcha(phone, function (json) {
            //倒计时
            that.codeCountDown();
        }, function (error: tomato.PError) {
            // funs.tip(error.name, error.note, "fail");
            if (that._flag) {
                that.find(".promptBar").removeClass('hide');
                that.find(".promptBar").text(error.note);
                setTimeout(function () {
                    that.find(".promptBar").addClass('hide');
                }, 3000);
            }

        });

    }


    /**登录 */
    _evt_toLogin() {
        var that = this;
        var phone = model.globalData.user.mobile;
        var code = this._code;
        var push_reg_id = "";//window['deviceData'].registerId;
        if (code) {
            api.Login(phone, code,push_reg_id, function (data: any) {
                that._call_back_function();
            }, function (error:tomato.PError) {
                funs.tip("验证码有误，请重新输入", "", "fail");
            });
        }
    }
    _afterInstallTo(){
        $(document).ready(function(){  $("input[name=focus]").focus();});
    }
}
export = VPresenter;
